controllers/cbcontainerscluster_controller.go

```diff
@@ -115,6 +159,13 @@ func (r *CBContainersClusterReconciler) Reconcile(ctx context.Context, req ctrl.
 	return ctrl.Result{Requeue: stateWasChanged}, nil
 }
 
+func (r *CBContainersClusterReconciler) apiGateway(cbContainersCluster *cbcontainersv1.CBContainersCluster, accessToken string) Gateway {
+	if r._apiGateway == nil {
+		r._apiGateway = r.GatewayCreator.CreateGateway(cbContainersCluster, accessToken)
+	}
+	return r._apiGateway
```

Comment:
I think you need a lock here if you want to treat `_apiGateway` as a singleton.
2 routines can run in this way, using 2 instances in practice

```
R1 runs line 163, enters if
R2 runs line 163, enters if
R1 runs line 164, assigns _apigateway
R1 runs line 166, returns _apigateway (1)
R2 runs line 164, assigns _apigateway again
R2 runs line 166, returns _apigateway(2)
```


On a separate note - what are the pros of this being a singleton - is it expensive to create ? 

As 1 con, this will essentially lock the instance to use the same access token indefinitely, which might not be good long-term if tokens need refreshing (e.g. if we move to OAuth or similar).

One idea could be to use 1 instance of the HTTP client but make the authentication per request - so that you create the expensive parts once but set the dynamic parts (e.g. Auth) every time. PSCKit's http client uses a similar approach and I think it is elegant.

Reply:
this singleton bring more harm than good. I removed it